pkg/controller/devicetainteviction/device_taint_eviction.go

```diff
@@ -0,0 +1,916 @@
+	pod, err := c.CoreV1().Pods(podRef.Namespace).Get(ctx, podRef.Name, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+
+	if pod.UID != podRef.UID {
+		// This special error suppresses event logging in our caller and prevents further retries.
+		// We can stop because the pod we were meant to evict is already gone and happens to
+		// be replaced by some other pod which reuses the same name.
+		return apierrors.NewNotFound(v1.SchemeGroupVersion.WithResource("pods").GroupResource(), pod.Name)
+	}
+
+	// Emit the event only once, and only if we are actually doing something.
+	if *emitEventFunc != nil {
+		(*emitEventFunc)(podRef)
+		*emitEventFunc = nil
+	}
+
+	newStatus := pod.Status.DeepCopy()
+	updated := apipod.UpdatePodCondition(newStatus, &v1.PodCondition{
+		Type:    v1.DisruptionTarget,
+		Status:  v1.ConditionTrue,
+		Reason:  "DeletionByDeviceTaintManager",
+		Message: "Device Taint manager: deleting due to NoExecute taint",
+	})
+	if updated {
+		if _, _, _, err := utilpod.PatchPodStatus(ctx, c, pod.Namespace, pod.Name, pod.UID, pod.Status, *newStatus); err != nil {
+			return err
+		}
+	}
+	// Unlikely, but it could happen that the pod we got above got replaced with
+	// another pod using the same name in the meantime. Include a precondition
+	// to prevent that race. This delete attempt then fails and the next one detects
+	// the new pod and stops retrying.
+	return c.CoreV1().Pods(podRef.Namespace).Delete(ctx, podRef.Name, metav1.DeleteOptions{
+		Preconditions: &metav1.Preconditions{
+			UID: &podRef.UID,
+		},
+	})
+}
+
+// New creates a new Controller that will use passed clientset to communicate with the API server.
+// Spawns no goroutines. That happens in Run.
+func New(c clientset.Interface, podInformer coreinformers.PodInformer, claimInformer resourceinformers.ResourceClaimInformer, sliceInformer resourceinformers.ResourceSliceInformer, taintInformer resourcealphainformers.DeviceTaintRuleInformer, classInformer resourceinformers.DeviceClassInformer, controllerName string) *Controller {
+	metrics.Register() // It would be nicer to pass the controller name here, but that probably would break generating https://kubernetes.io/docs/reference/instrumentation/metrics.
+
+	tc := &Controller{
+		name: controllerName,
+
+		client:          c,
+		podInformer:     podInformer,
+		podLister:       podInformer.Lister(),
+		claimInformer:   claimInformer,
+		sliceInformer:   sliceInformer,
+		taintInformer:   taintInformer,
+		classInformer:   classInformer,
+		allocatedClaims: make(map[types.NamespacedName]allocatedClaim),
+		pools:           make(map[poolID]pool),
+		// Instantiate all informers now to ensure that they get started.
+		haveSynced: []cache.InformerSynced{
+			podInformer.Informer().HasSynced,
+			claimInformer.Informer().HasSynced,
+			sliceInformer.Informer().HasSynced,
+			taintInformer.Informer().HasSynced,
+			classInformer.Informer().HasSynced,
+		},
+		metrics: metrics.Global,
```

Comment:
What's the reason why you need to use a global? Couldn't you call `metrics.New()` here and then modify the `Register` to take Metrics as receiver?

Reply:
This is the constructor called by kube-controller-manager. kube-controller-manager expects all controllers to use global metrics. I didn't want to break that contract.

Reply:
For example, while unusual it wouldn't be wrong to call New multiple times for the same name as long as only one of them gets to run or the running instance is stopped.

Reply:
I tried, but gave up because it also broke unit testing where New is called multiple times and the metrics names are expected to use the same name for all test cases.